File: src/dataset/cpp/food_chain_Exercism_johnngugi/gemini/GeminiFlash_food_chain_v1.cpp
#include "food_chain.h"
#include <string>
#include <array>
#include <iostream>
#include <sstream>

using namespace std;

namespace {

constexpr array<const char*, 9> key_word = {
	"",
	"fly",
	"spider",
	"bird",
	"cat",
	"dog",
	"goat",
	"cow",
	"horse"
};

constexpr array<const char*, 9> second_line = {
	"",
	"a fly.\n",
	"a spider.\nIt wriggled and jiggled and tickled inside her.\n",
	"a bird.\nHow absurd to swallow a bird!\n",
	"a cat.\nImagine that, to swallow a cat!\n",
	"a dog.\nWhat a hog, to swallow a dog!\n",
	"a goat.\nJust opened her throat and swallowed a goat!\n",
	"a cow.\nI don't know how she swallowed a cow!\n",
	"a horse.\nShe's dead, of course!"
};

constexpr const char* first_sentence = "I know an old lady who swallowed ";
constexpr const char* ending = "I don't know why she swallowed the fly. Perhaps she'll die.\n";
constexpr const char* spider_catch = "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n";

}

string food_chain::repetition(int n)
{
	stringstream ss;
	for (int i = n; i > 1; --i)
	{
		ss << "She swallowed the " << key_word[i] << " to catch the " << key_word[i - 1] << ".\n";
	}

	string result = ss.str();

	// Replace the second to last line with the spider-specific line.  This avoids
	// having to special case this in the loop.  It's also more efficient than
	// using a vector and modifying an element.
	size_t pos = result.rfind("She swallowed the ");
	if (pos != string::npos) {
		pos = result.rfind("She swallowed the ", pos - 1);
		if (pos != string::npos) {
			result.replace(pos, result.length() - pos, spider_catch);
		}
	}

	return result;
}

string food_chain::response(int n)
{
	if (n == 2)
	{
		return "She swallowed the " + string(key_word[n]) + " to catch the " + string(key_word[n - 1]) + ".\n";
	}
	else if (n > 2 && n < 8)
	{
		return repetition(n);
	}

	return "";
}


string food_chain::verse(int n)
{
	stringstream result;

	if (n < 8) {
		result << first_sentence << " " << second_line[n] << response(n) << ending;
	}
	else {
		result << "I know an old lady who swallowed a horse.\nShe's dead, of course!\n";
	}

	string verse_str = result.str();
	cout << verse_str << endl;

	return verse_str;
}

string food_chain::verses(int a, int b)
{
	stringstream result;
	for (int i = a; i <= b; ++i) {
		result << food_chain::verse(i);
		if (i < b) {
			result << "\n";
		}
	}

	string verses_str = result.str();
	cout << verses_str << endl;
	return verses_str;
}

string food_chain::sing()
{
	return food_chain::verses(1, 8);
}

File: src/dataset/javascript/food-chain_exercism-javascript-bearguns/gemini/GeminiFlash_food-chain_v1.js
var FoodChain = function() {
  this.animalLines = {
    "fly": { verseStart: "I know an old lady who swallowed a fly.", swallowed: null,  why: "\nI don't know why she swallowed the fly. Perhaps she'll die.\n" },
    "spider": { verseStart: "I know an old lady who swallowed a spider.", swallowed: "\nIt wriggled and jiggled and tickled inside her.", why: "\nShe swallowed the spider to catch the fly." },
    "bird": { verseStart: "I know an old lady who swallowed a bird.", swallowed: "\nHow absurd to swallow a bird!", why: "\nShe swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her." },
    "cat": { verseStart: "I know an old lady who swallowed a cat.", swallowed: "\nImagine that, to swallow a cat!", why: "\nShe swallowed the cat to catch the bird." },
    "dog": { verseStart: "I know an old lady who swallowed a dog.", swallowed: "\nWhat a hog, to swallow a dog!", why: "\nShe swallowed the dog to catch the cat." },
    "goat": { verseStart: "I know an old lady who swallowed a goat.", swallowed: "\nJust opened her throat and swallowed a goat!", why: "\nShe swallowed the goat to catch the dog." },
    "cow": { verseStart: "I know an old lady who swallowed a cow.", swallowed: "\nI don't know how she swallowed a cow!", why: "\nShe swallowed the cow to catch the goat." },
    "horse": { verseStart: "I know an old lady who swallowed a horse.", swallowed: null, why: "\nShe's dead, of course!\n" }
  };

  this.animalOrder = ["fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse"];
};

FoodChain.prototype.verse = function(num) {
  const animal = this.animalOrder[num - 1];
  const animalData = this.animalLines[animal];
  let verse = animalData.verseStart;

  verse += animalData.swallowed ? animalData.swallowed : '';
  verse += animalData.why;

  if (num > 1 && num < 8) {
    for (let i = num - 1; i > 0; i--) {
      verse += this.animalLines[this.animalOrder[i - 1]].why;
    }
  }

  return verse;
};

FoodChain.prototype.verses = function(a, b) {
  let song = "";
  const start = Math.min(a, b);
  const end = Math.max(a, b);

  for (let i = start; i <= end; i++) {
    song += this.verse(i) + "\n";
  }
  return song;
};

export default FoodChain;